integrated circuits - assemblies are now fun to smash and destroy

Tazaraki opened the pull request:
:cl: Taza, gy1ta23
tweak: integrated circuits - destroying circuit assemblies now features sparks, on-hit effects, and debris 
tweak: integrated circuits - assemblies have had their descriptions, chat notices, and health values updated
/:cl:

I originally thought circuit assemblies were indestructible. Turns out they weren't, but it was not obvious when they were being hurt or destroyed. Barely anything would happen in-game and then they would disappear. 

This is my go at making it more enjoyable to smack these things with crowbars or shoot them with guns. I'm including [gy1ta23](https://github.com/gy1ta23) in the change log credit for being very helpful in finding the references I needed to make this work. 

Here is a comprehensive list of changes:

General changes:
- All assembly descriptions have been updated
- Many outdated span tags have been updated 
- New chat notifications have been added when damaging an assembly and observing a damaged assembly
- Large and Drone type assemblies now consistently center themselves on tiles when dropped

Circuit assemblies now:
- have new health values matching their size and design
- have an on-hit cooldown for all interactions and an on-hit animation when being smashed
- play different on-hit sounds depending on if they were struck, shot with a bullet, or shot with a laser
- have their screws become loose if they take too much damage
- spark if struck when brought beneath a specific health threshold
- spark periodically when left damaged under a specific health threshold
- play a sound when destroyed
- leave a varying amount of steel when destroyed depending on size, the drones additionally dropping robot gibs

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +52,46 @@
 
 
-/obj/item/device/electronic_assembly/proc/take_damage(var/amnt)
+/obj/item/device/electronic_assembly/proc/take_damage(var/amnt, var/asize)
```
`var/` is redundant in proc parameters.

```suggestion
/obj/item/device/electronic_assembly/proc/take_damage(var/amnt, asize)
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +52,46 @@
+
 	if(health <= 0)
-		visible_message("<span class='danger'>\The [src] falls to pieces!</span>")
+		visible_message(SPAN_WARNING("\The [src] falls to pieces!"))
+		if(asize == ITEM_SIZE_HUGE)
+			new /obj/effect/decal/cleanable/blood/gibs/robot(loc)
+			new /obj/item/stack/material/steel(src.loc, rand(7, 10))
+		else if(asize == ITEM_SIZE_LARGE)
+			new /obj/item/stack/material/steel(src.loc, rand(3, 6))
+		else if(asize == ITEM_SIZE_NORMAL)
+			new /obj/item/stack/material/steel(src.loc, rand(1, 3))
+		else
+			new /obj/item/stack/material/steel(src.loc)
+		playsound(src.loc, 'sound/items/electronic_assembly_empty.ogg', 100, 1)
+		new /obj/effect/sparks(src.loc, 1)
```
`src` is redundant when referencing the base path's vars, unless there's a locally scoped var of the same name.

```suggestion
		visible_message(SPAN_WARNING("\The [src] falls to pieces!"))
		if(asize == ITEM_SIZE_HUGE)
			new /obj/effect/decal/cleanable/blood/gibs/robot(loc)
			new /obj/item/stack/material/steel(loc, rand(7, 10))
		else if(asize == ITEM_SIZE_LARGE)
			new /obj/item/stack/material/steel(loc, rand(3, 6))
		else if(asize == ITEM_SIZE_NORMAL)
			new /obj/item/stack/material/steel(loc, rand(1, 3))
		else
			new /obj/item/stack/material/steel(loc)
		playsound(loc, 'sound/items/electronic_assembly_empty.ogg', 100, 1)
		new /obj/effect/sparks(loc, 1)
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +52,46 @@
-
+	else if(health <= initial(health)*0.25)
+		visible_message(SPAN_WARNING("\The [src] sputters and sparks!"))
+		new /obj/effect/sparks(src.loc, 1)
```
```suggestion
		new /obj/effect/sparks(loc, 1)
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -109,8 +128,9 @@
-		visible_message("<span class='warning'>\The [src] shudders and sparks</span>")
+	if(health <= initial(health)*0.25 && prob(5))
+		visible_message(SPAN_WARNING("\The [src] sparks violently!"))
+		new /obj/effect/sparks(src.loc, 1)
```
```suggestion
		new /obj/effect/sparks(loc, 1)
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -464,21 +485,23 @@
-			to_chat(user, "<span class='danger'>\The [user] hits \the [src] with \the [I]</span>")
+		if(user.a_intent == I_HURT && (!(user.l_hand == src || user.r_hand == src))) // Kill it
+			user.do_attack_animation(src)
+			playsound(src.loc, 'sound/weapons/genhit1.ogg', 100, 1)
```
```suggestion
			playsound(loc, 'sound/weapons/genhit1.ogg', 100, 1)
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -489,6 +512,10 @@
 
 /obj/item/device/electronic_assembly/bullet_act(var/obj/item/projectile/P)
 	take_damage(P.damage)
+	if(istype(P,/obj/item/projectile/beam))
+		playsound(src.loc, SOUNDS_LASER_METAL, 100, 1)
+	else if(istype(P,/obj/item/projectile/bullet))
+		playsound(src.loc, SOUNDS_BULLET_METAL, 100, 1)
```
```suggestion
	if(istype(P,/obj/item/projectile/beam))
		playsound(loc, SOUNDS_LASER_METAL, 100, 1)
	else if(istype(P,/obj/item/projectile/bullet))
		playsound(loc, SOUNDS_BULLET_METAL, 100, 1)
```

Tazaraki commented:
I have removed the redundancies that were listed.

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +52,46 @@
-/obj/item/device/electronic_assembly/proc/take_damage(var/amnt)
+/obj/item/device/electronic_assembly/proc/take_damage(var/amnt, asize)
 	health = health - amnt
+	asize = w_class
```
What's the purpose of the `asize` parameter if you're immediately changing its value? Also, if `asize` is just going to be `w_class` without modifying it in any way in this proc, why not just use `w_class` below?

Tazaraki replied:
Good question. I believe I was messing with asize being something other than just w_class, and then settled on this. I've removed asize and replaced it with w_class in the code.

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +52,46 @@
+			new /obj/item/stack/material/steel(loc, rand(1, 3))
+		else
+			new /obj/item/stack/material/steel(loc)
+		playsound(src.loc, 'sound/items/electronic_assembly_empty.ogg', 100, 1)
```
I missed one.

```suggestion
		playsound(loc, 'sound/items/electronic_assembly_empty.ogg', 100, 1)
```

Tazaraki replied:
I did what I should've done earlier and ctrl+f'd all instances of src.loc and removed the remaining three.

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +52,45 @@
-
+	else if(health <= initial(health)*0.25)
+		visible_message(SPAN_WARNING("\The [src] sputters and sparks!"))
+		new /obj/effect/sparks(loc, 1)
```
I've learned that calling the spark effects directly isn't a good idea. You should instead use the spark spread datum: `/datum/effect/effect/system/spark_spread`

You'll want to add a var to `/obj/item/device/electronic_assembly` that is of type `/datum/effect/effect/system/spark_spread` and on Initialize creates a new instance of itself (i.e. `spark_system= new spark_system()`), then call the `set_up()` and `attach()` procs. Here's an example of the setup calls from turrets:

```dm
	//Sets up a spark system
	spark_system = new /datum/effect/effect/system/spark_spread
	spark_system.set_up(5, 0, src)
	spark_system.attach(src)
```

You'll also need to update the Destroy() proc to `qdel_null(spark_system)`

Then, when you want to actually generate sparks, run `spark_system.start()` to generate the sparks.

Tazaraki replied:
Thanks a ton for providing this guide and example. You've led me to study the effect_system.dm further to understand why this is a better idea and I think I've learned a lot. I just pushed some changes.

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -26,6 +26,8 @@
 	var/static/next_assembly_id = 0
 	var/interact_page = 0
 	var/components_per_page = 5
+	var/datum/effect/effect/system/spark_spread/spark_system	// Spark system used for creating sparks while damaged
```
Use dmdocs if you're going to be commenting properties - https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks

```suggestion
	/// Spark system used for creating sparks while damaged
	var/datum/effect/effect/system/spark_spread/spark_system
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -26,6 +26,8 @@
 	var/interact_page = 0
 	var/components_per_page = 5
+	var/datum/effect/effect/system/spark_spread/spark_system	// Spark system used for creating sparks while damaged
+	var/adrone = 0
```
This looks like it's intended to be a true/false property - If so, use the `TRUE` and `FALSE` defines.
```suggestion
	var/adrone = FALSE
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +54,49 @@
+		else if((health <= initial(health)*0.75))
+			to_chat(user, SPAN_DANGER("\The [src] is covered in dents and punctured in several places."))
+		else	
+			to_chat(user, SPAN_NOTICE("\The [src] has a few scuffs and scratches."))
```
Another dev mentioned this can be condensed to a switch.
```suggestion
		switch (health / initial(health))
			if (0.75 to INFINITY)
				to_chat(user, SPAN_NOTICE("\The [src] has a few scuffs and scratches."))
			if (0.5 to 0.75)
				to_chat(user, SPAN_DANGER("\The [src] is covered in dents and punctured in several places."))
			if (0.25 to 0.5)
				to_chat(user, SPAN_DANGER("\The [src] looks seriously damaged!"))
			else
				to_chat(user, SPAN_WARNING("\The [src] is barely holding together!"))
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +54,49 @@
+			new /obj/effect/decal/cleanable/blood/gibs/robot(loc)
+			new /obj/item/stack/material/steel(loc, rand(7, 10))
+		else if(w_class == ITEM_SIZE_LARGE)
+			if(adrone == 1)
```
`== 1` and `== TRUE` are rundant for TRUE/FALSE vars.

```suggestion
			if(adrone)
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -52,27 +54,49 @@
+		spark_system.start()
+		playsound(loc, 'sound/items/electronic_assembly_empty.ogg', 100, 1)
+		icon = 0
+		sleep(6)
```
Don't add new sleeps if you can avoid it. Use timers instead with the `addtimer()` proc and a new proc you add to the object as the handler for the timer. The timer will call the referenced proc when it completes, and won't cause the existing proc-chain to be delayed by the timer. Here's an example from another file:

```dm
/obj/effect/gateway/active/New()
	..()
	addtimer(CALLBACK(src, .proc/create_and_delete), rand(30,60) SECONDS)


/obj/effect/gateway/active/proc/create_and_delete()
	var/t = pick(spawnable)
	new t(src.loc)
	qdel(src)
```

SierraKomodo commented on code/modules/integrated_electronics/core/assemblies.dm:
```diff
@@ -602,55 +638,58 @@
-	health = 50
+	health = 60
+	randpixel = 0
+	adrone = 1
```
```suggestion
	adrone = TRUE
```

Tazaraki committed:
circuits are now fun to smash and destroy


adjusts damaged spark rate and updates some span tags


changed variable dronesize to asize


removed redundancies in take_damage() and playsounds 


removed asize variable, removed more src.loc redundancies


creates a spark system and applies it to the assembly


fixed a small dreamchecker error


missed an equal sign, oops


optimized a few areas of code and made sparks dependent on batteries

Tazaraki commented:
I've studied up on the changes suggested and did my best to implement them. I also added a few lines of code that make sparks dependent on a battery being in the device and the battery having charge.
